Submissions

gingin77 opened the pull request:
**What:** The Submission class handles both posts and comments. A comment is a submission that has a parent_post_id, whereas the parent_post_id is nil for posts. There are separate forms and create methods for each. The advantage to this approach is that one table holds all submissions.

**Hoping for feedback on**
	_Ways to refactor the model and controller_
            - Are there better places to put certain chunks of code (like the `limit_text method`)? 
            - Are there existing Rails helpers or Ruby methods available that come to mind, which could replace any of the code I’ve written?
            - Are there better ways to write the queries I have in the Submission model?
	- The views seem repetitive, but I haven’t tried to use partials because nothing has jumped out as something easy to modularize. I’ve read about the content_for method and I'm wondering if that could be useful.

**Next steps** 
	- Rewrite tests for Submission model in Rspec
	- Implement a reply option after a comment is posted - a user comments on a post and replies to existing comments. The replies will then be displayed as nested underneath the ‘parent’ comment
	- Model votes for both posts and comments

gingin77 committed:
Create Submission model and controller with basic functionality, though validations are needed.

Anyone can view a list of submissions or a single submission
Authenticated users can submit a new submission, though the type of submission needs to be refined to either title & URL OR text.

gingin77 committed:
WIP - Build some tests and validation code in the model

gingin77 committed:
Update fixtures to address errors and switch version of minitest gem

See issue #730 on https://github.com/seattlerb/minitest/issues/730 - I used one of the suggested workarounds and switched the minitest version in the gemfile

The fixtures needed id's because submissions belong to users

gingin77 committed:
Write tests for Submissions model and setup validations - all subm. model tests pass

gingin77 committed:
Refactor submission_test and model

Update tests so objects being used in test are contained within each test instead of a setup method
Remove fixture files
In model, update custom validations to place message on error base, not :text

gingin77 committed:
WIP - further refactoring of the tests for the Submission model

gingin77 committed:
Refactor submissions model and tests from top

Add validation on user_id
Simplify how a url_sub is validated
Simplify how a text_sub is validated
Add validation to require at least one field be present

gingin77 committed:
Add null constraint on users in submissions table; modify tests

gingin77 committed:
Add FactoryBot

gingin77 committed:
Abbreviate string in test for url

gingin77 committed:
Add relationship to subs in user model

gingin77 committed:
Set route root to submissions#index

gingin77 committed:
Add seed data to populate sub index view; Add details to improve Sub#index view

View edit that goes with commit f4108b7

gingin77 committed:
Remove fields that for notice and alert messages from layouts view

gingin77 committed:
In sub test - use build(:user) instead of create

gingin77 committed:
Set up controller and views to require authentication for new submissions

Use :alert, instead of danger; modify styles so default bootstrap styles are not so obvious looking

gingin77 committed:
Update div wrappers to style edit and show views

gingin77 committed:
Define limit_text method and apply it in the index view

gingin77 committed:
Apply more ruby to views to improve UX

gingin77 committed:
Set up post-comments relationship within Sub model

Modify original Submissions table migration to include user null constraint and new reference to post. Delete migration that solely added null constraint

gingin77 committed:
New comments can be submitted via submission#show page

Within Submission class - posts and comments are defined within controller as instance vars. Posts are submissions where the post_id is nil, meaning a post does NOT have a parent submission, where as a comment is created using the parent submission it is associated with

gingin77 committed:
Refactor controller action methods, helpers, and sequence; modify views to include corrent helper names

gingin77 committed:
Change references to user instead of profile, after rebasing user_model branch into this submissions feature branch

nhunzaker commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,77 @@
+  end
+
+  def new_comment
+    @comment = Submission.new
```
Should these be cached, like `@comment ||= Submission.new`, so they do not create new records on every call?

gingin77 replied:
Thanks @nhunzaker, I've been stuck on this because I thought the point of `.new` is that a new object will be instantiated but it's not saved with AR until `.save` is applied. Still, I guess that new object is still taking up space in the app's working memory.

nhunzaker commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,77 @@
+  before_action :authenticate_user!, only: [:create, :create_comment]
+  before_action :authenticate_to_submit, only: :new
+
+  helper_method :posts, :post, :new_post, :new_comment
```
Oooh neat. I didn't know this was a thing.

gingin77 replied:
Do you mean listing the helper_methods at the top after the before filters? I thought they'd be easier to quickly see there.... though some people may look for `helper_method :posts` right after the private method.

nhunzaker replied:
Yep, `:helper_method`. I've never seen it used before.

nhunzaker replied:
And I think it makes sense up here, where it's really easy to catch when scanning.

nhunzaker commented on app/models/submission.rb:
```diff
@@ -0,0 +1,40 @@
+      break if shift_pos > 0
+    end
+    "#{ary.shift(ary.count - shift_pos).join}..."
+  end
```
Could https://apidock.com/rails/String/truncate work?

gingin77 replied:
Ah, of course a truncate method exists - I never looked! But, I like the way my method cuts the string on a blank space between words. I'll keep it but glad to know about truncate for future projects.

nhunzaker replied:
Ah cool. I wonder if the `separate` option gets you something similar. It looks like you can do something like:

```ruby
"The quick brown fox jumps over the fence".truncate(25, separator: /\s/)
# "The quick brown fox..."
```

gingin77 replied:
Perfect! I'd glanced at that link too quickly. 

Man, I need to stop spending time writing methods that Ruby and Rails already provide! At least now I'm inspired to pick up more new methods from the source code.

nhunzaker commented on app/models/submission.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+
+    def text_submission
+      if text.present? && (url.present? || title.present?)
```
No action required: I just learned that `.present?` returns `false` on sparsely empty strings (like `'   '`). Neat.

gingin77 replied:
I learned that from Pat last week.

zporter replied:
Can use the `url_submission?` method here?

gingin77 replied:
Technically, yes, but I feel like that would make the code less human readable.

Actually, I don't like how I currently define the 2 submission types - with one being a boolean and the other adding an error message. It seems to obscure the either one or the other relationship.

gingin77 replied:
Following my last comment - here's a refactor https://github.com/gingin77/hackernews_clone/pull/5/commits/009139844e27e461bcdd49c4af25fd92a2a494ea

nhunzaker commented on app/models/submission.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+
+    def at_least_one_field_present
+      if text.blank? && url.blank? && title.blank?
```
Why `blank?` here, but `present?` above?

gingin77 replied:
Good question - the method above, `text_submission`, is designed to say, if you submit a text post, then the url and title fields should NOT be present, so a message is added to the errors object. This method, `at_least_one_field_present` is saying, if all of the fields are blank, add to the errors object. 

Do you think the following would be more readable for `text_submission`? 
`if text.present? && (!url.blank? || !title.blank?)`

nhunzaker replied:
Sorry. I left a bad comment. `blank?` is the opposite of `.present?`. Sorry :(

gingin77 replied:
no worries

zporter replied:
Fun Ruby trick: `if [text, url, title].all?(&:blank?)`

Might be too clever, but still fun nonetheless.

nhunzaker commented on config/routes.rb:
```diff
@@ -1,7 +1,10 @@
   resources :users, only: :show
+
+  resources :submissions do
+    post 'create_comment'
```
In the past, I've seen create actions just as expressed `post '<resource>'`, like `post 'comment'`. I want to say this follows REST more closely, but I don't know if there's an existing Rails convention.

What do you think about removing the `create_` here?

zporter replied:
I think that's a great idea, @nhunzaker, but I can also see a couple other options here:

1. Modify `submissions#create` to handle both posts **and** comments.
1. Create a new `comments#create` action to handle comment submissions.

nhunzaker commented on db/seeds.rb:
```diff
@@ -1,12 +1,29 @@
+Submission.create([
+  {
+    user_id: 1,
+    text: "Plants don’t get enough credit. They move. You know this. Your houseplant salutes the sun each morning. At night, it returns to center."
```
So true! I have to rotate my little stemmy succulent or it gets all bendy following the light.

nhunzaker commented on test/factories/submission.rb:
```diff
@@ -0,0 +1,18 @@
+FactoryBot.define do
+  factory :user do
+    name "Oliver Sacks"
```
Personal favorite.

gingin77 replied:
Yeah, somehow lots of links on the HN site are from science writers!

nhunzaker commented on app/assets/stylesheets/application.scss:
```diff
@@ -68,3 +80,20 @@ a {
+
+.alert {
+    margin-top: 1rem;
+}
```
4 spaces here, but 2 in other places.

nhunzaker commented on db/schema.rb:
```diff
@@ -10,11 +10,23 @@
+    t.string "url"
+    t.text "text"
+    t.bigint "user_id", null: false
+    t.bigint "post_id"
```
Does a null here map to a submission that just asks a question to the community?

gingin77 replied:
The null here is operating at the data table level to require that a `user_id` be present.

Are you asking if a user can post a question to the community to start a discussion thread? Someone could use the the "text" post option for that. Maybe no one makes that connection and that's why there are so few "text" posts on HN.

nhunzaker replied:
Ah I was referring to the `post_id`. Cool. So a `text` type post has no post. 👍

zporter replied:
@nhunzaker -- yes, the idea is that if `post_id` is `null`, then a submission should be treated as a "post" vs. a "comment".

gingin77 replied:
@nhunzaker - sorry for not getting what you were asking. Part of the confusion may be because I wrote validations for Submissions with posts in mind, then added the comment functionality, but haven't gone back to update the validations.

Here's a break down of the options:
**A Submission can be a post OR a comment**
 * If post_id is null, the Submission is a post;
 * If post_id is not null, the Submission is a comment

**When submitting a post, a User can submit either**
 * a url and title OR, 
 * text content

**When submitting a comment, a user submits**
* text content

*Functionality that hasn't yet been written*
**Comments can be:**
* In response to posts
* Replies to other comments

As I wrote in the PR description, the advantage to this approach is that one table holds all submissions. There could be disadvantages but I haven’t run into any . . . yet :)

nhunzaker reviewed (approved):
Couple of small things, but this looks good overall. I wish I could speak more to the database decisions. I'm looking forward to other's comments!

zporter commented on Gemfile:
```diff
@@ -27,17 +27,21 @@ gem 'turbolinks', '~> 5'
 gem 'bootstrap', '~> 4.0.0'
 
+
+
```
nitpick: 👾

gingin77 replied:
Hmm, I see a few possible issues here. I'm not sure which thing you're point out. Should I address all bullet points below?

- I didn't include a version # for devise... I can
- there are lines above and below this code block. Should I stack all the gems without any line spacing between - except to distinguish the groups at the end?

Also, there are a few gems present that I'm not sure if I need like coffee-rails and jbuilder. 
- I've only used jbuilder when building an API. Does it have other functions?
- As for minitest - I'm planning to phase that out soon, but the existing tests need it

zporter replied:
I was pointing out that there was already a blank line and 2 more were added in this commit. I'm not sure 3 blank lines are necessary for separating logical / grouped bits of code.

> I didn't include a version # for devise

Not necessary because the version can be found in the `Gemfile.lock` file, though it wouldn't hurt to specify it here so a `bundle update` wouldn't accidentally upgrade it and break all the things.

> there are lines above and below this code block. Should I stack all the gems without any line spacing between - except to distinguish the groups at the end?

I was pointing the spacing out yes. I don't think you need to remove all of them, but I also think 3 is a tad excessive. But totally a stylistic choice, so that's why I prefaced with `nitpick`.

> I've only used jbuilder when building an API. Does it have other functions?

That's essentially what it's good for -- for defining JSON responses. Since we're not building that (yet), I'd say you're fine removing it.

Hope that helps to clarify the meaning of 👾 emoji!

zporter commented on app/assets/stylesheets/application.scss:
```diff
@@ -68,3 +80,20 @@ a {
+.alert-alert, .alert-notice {
+    color: #170838;
+    background-color: #ff590080;
+    border-color: #ff590080;
```
looks like there's additional indentation here compared to other CSS rules above.

zporter commented on app/assets/stylesheets/application.scss:
```diff
@@ -68,3 +80,20 @@ a {
+}
+
+.alert {
+    margin-top: 1rem;
```
looks like there's additional indentation here compared to other CSS rules above.

zporter commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,77 @@
+      redirect_to submission_path(@post.id)
+    else
+      flash[:alert] = @post.errors.messages
+      redirect_to new_submission_path
```
I'd rather see `render :new` used here so that the data the user submitted is persisted on the form. As a poster, I'd hate to write a whole text blurb and have that lost because of a validation error.  The flash notification would need to be updated to use `flash.now` in order to continue to work.

zporter commented on db/migrate/20180131214527_create_submissions.rb:
```diff
@@ -0,0 +1,12 @@
+      t.belongs_to :user, index: true, null: false
+      t.references :post, index: true
+      t.timestamps
+    end
```
I think this is a good start to simply treating "comments" as text posts on another submission. Let's say a feature request comes in that says "I want to show the total number of comments next to a post" (not unlike the "66 comments" in the following screenshot).

![screen shot 2018-02-06 at 9 01 34 am](https://user-images.githubusercontent.com/250031/35863356-6af91354-0b1c-11e8-97e1-a0e6f708837e.png)

How would you handle that with this data model?  Keep in mind that comments are nestable, meaning that a comment can have many comments.

gingin77 replied:
My first thought is to add a column for comment_id, similar to how I set up post_id. If comment_id is nil, then it is a comment directly in response to the post; if the comment_id has a number, it is a nested comment in reply to a parent comment. 

I'd also like to have an index that shows the sequence of the nested comments - as in first, second, third, or etc nested comment (indicated by an index value), under parent comment (indicated by comment_id). I've not used an index in this way before so I'm not sure if it can be executed the way I'm thinking.

If I can set up an index to indicate the sequence of nested comments relative to a parent comment, it would be good to also set up an index that tracks the sequence of non-nested comments relative to a parent post.

zporter replied:
> My first thought is to add a column for comment_id, similar to how I set up post_id. If comment_id is nil, then it is a comment directly in response to the post; if the comment_id has a number, it is a nested comment in reply to a parent comment.

That's a good thought, but I'm thinking of how would the view look like if you just want to iterate through immediate comments instead of all comments on a post (including nested)?  Probably with some sort of scope like the following?

```ruby
class Submission < ApplicationRecord
  has_many :direct_comments,
           ~> { where(comment_id: nil) },
           class_name: 'Submission',
           foreign_key: :post_id
end
```

Then in the view, you'd loop over them like:

```erb
<%= post.direct_comments.each do |comment| %>
  <%= render comment %>
<% end %>
```

That relationship and scope only works for the most-parent submission (a Post), but breaks down on comment submissions.

My thought on this would be to rename the `post_id` column to a more generic `parent_id` with a foreign key constraint setup on `submissions.id`. Then add a `counter_cache` column to keep track of the number of comments submitted to a post. You can read more about counter caches [here](http://guides.rubyonrails.org/association_basics.html) (search for `4.1.2.3 :counter_cache`). If you still want to proceed forward with the `post_id` and `comment_id` columns, then that says to me that these should be 2 separate tables (`posts` and `comments`) to better encapsulate the business logic.

> I'd also like to have an index that shows the sequence of the nested comments - as in first, second, third, or etc nested comment (indicated by an index value), under parent comment (indicated by comment_id). I've not used an index in this way before so I'm not sure if it can be executed the way I'm thinking.

> If I can set up an index to indicate the sequence of nested comments relative to a parent comment, it would be good to also set up an index that tracks the sequence of non-nested comments relative to a parent post.

That sounds cool. I've handled this in the past with a simple `position` column on the table. You can either pull in a gem, write a PostgreSQL sequence, or roll your own Ruby code to handle incrementing and updating that column as comments are added / removed.  Honestly though, the simplest approach would be to sort by the `created_at` timestamp with possibly the most recent first. Once voting gets into place, perhaps we'll consider sorting by "most upvoted".

Hope that helps!

zporter commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,77 @@
+  end
+
+  def posts
+    @posts = Submission.where(post_id: nil)
```
Use `||=` here to memoize the AR call.

Also, might want to consider adding a scope on Submission here if you find this query getting used over and over. Something like:

```ruby
class Submission < ApplicationRecord
  # ... other bits omitted ...

  # use `Submission.posts` to nab posts only
  scope :posts, ~> { where post_id: nil }

  # ... other bits omitted ...
end
```

you can find more info in the [docs](http://api.rubyonrails.org/classes/ActiveRecord/Scoping/Named/ClassMethods.html#method-i-scope)

zporter commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,77 @@
+
+  def new_post
+    @post = Submission.new
+  end
```
I'd probably drop this in favor of using the `post` method across all actions. Simply set `@post = current_user.build_submission` in the `new` method above and any call to `post` will automatically pull from that ivar.

zporter commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,77 @@
+  end
+
+  def post
+    @post = Submission.find(params[:id])
```
Use `||=` here to memoize the value under the `@post` instance variable.

zporter commented on app/controllers/submissions_controller.rb:
```diff
@@ -0,0 +1,77 @@
+  def create_comment
+    @comment = Submission.new(comment_params)
+    @comment.user_id = current_user.id
+    @comment.post_id = parent_post_id
```
instead of setting the attributes, the params hash could be modified to include them. I'm thinking something like:

```ruby
@comment = current_user.submissions.build(comment_params.merge(post_id: parent_post_id))
```

Could also modify the `comment_params` method to merge in the `post_id`. Your call.

gingin77 replied:
Nice - I forgot about merge, I was trying to use `(comments_params, {user_id: ..., etc })` and getting a too many arguments error

gingin77 replied:
Also, this helps resolve some confusion I've been having about the relationship between the `new_comment` method and this `create_comment`method.... I think you may have made the issue explicit in your comment on the `new_post` method.

I needed the `new_post` and `new_comment` methods to support the blank form fields on the GET, but the #create actions also have `Submission.new`. I didn't think I needed `Submission.new` twice per post or comment, but wasn't sure where to have it.

gingin77 committed:
Update indentation on styles

zporter commented on app/views/layouts/application.html.erb:
```diff
@@ -13,7 +13,8 @@
       <div class="nav_always_show">
         <%= image_tag "y18.gif" %>
-        <span>HackerNewsClone</span>
+        <span><%= link_to "HackerNewsClone", submissions_path %></span>
```
should this be to the `root_path` since it's the site logo?

zporter commented on app/views/layouts/application.html.erb:
```diff
@@ -23,8 +24,15 @@
+      <% elsif msg.kind_of?(String) %>
+        <%= content_tag(:div, msg, class: "alert alert-#{type}") %>
+      <% end %>
+    <% end %>
```
if you use `resource.errors.full_messages.to_sentence`, then a String is always returned and you won't need to worry about handling a Hash. but this is cool too if preferable.

gingin77 replied:
![screen shot 2018-02-06 at 3 19 52 pm](https://user-images.githubusercontent.com/30031865/35882180-4c4ac968-0b51-11e8-8ffa-044f2ad5eb60.png)

^^Just cuz its fun to add screenshots.... When I take out the code related to pulling messages from a Hash, I see the whole errors object: {:url=>["can't be blank"], :base=>[], :title=>[], :text=>[]}

I'm taking out the Hash vs String syntax here.  That was the solution I implemented in grade book after seeing a similar error message in that app.

I’m working on a set up where field-specific messages will show up inline, just below the missing element. If the error is on base, it should show up at just above the top of the form.

zporter replied:
> When I take out the code related to pulling messages from a Hash, I see the whole errors object

Right, because it's calling `to_s` on `resource.errors`.  If you nab `full_messages` off of that, you should see an array of messages like `["Url can't be blank", "Title can't be blank"]`.  Then, add `to_sentence` to that and get a string like `"Url can't be blank and Title can't be blank"`. So the full code is `resource.errors.full_messages.to_sentence`.  Replacing `resource` with... `submission`(?) in this instance.

I like the idea of the inline errors though 👍

zporter commented on app/views/submissions/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+<% posts.map do |s| %>
```
Use of `map` is interesting here. Did you mean to use `each`?

zporter replied:
I'd also rename `s` to `post` to be a bit more descriptive about what the current iteration is.

zporter commented on app/views/submissions/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+<% posts.map do |s| %>
+  <br />
+  <% if s.text.blank? %>
```
use the `url_submission?` predicate method?

zporter commented on app/views/submissions/index.html.erb:
```diff
@@ -0,0 +1,11 @@
+  <% if s.text.blank? %>
+    <%= link_to(s.title, s.url) %>
+  <% else %>
+    <%= s.limit_text %>
```
I'm beginning to wonder if it'd be a better user experience if this were linked to the `submissions#show` view. 🤔

gingin77 replied:
I've been thinking that too. What do you think of requiring a title on text posts?

zporter replied:
I can dig it. Are you envisioning limiting a Title's length?

zporter commented on app/views/submissions/show.html.erb:
```diff
@@ -0,0 +1,38 @@
+<div class="show_and_edit">
+  <% if post.text.blank? %>
```
use the `url_submission?` predicate method?

zporter commented on app/models/submission.rb:
```diff
@@ -0,0 +1,40 @@
+class Submission < ApplicationRecord
+  belongs_to :user
```
thoughts on referring to this relationship as an `owner`, `author`, `submitter`, or a `poster`?

gingin77 replied:
If its possible to set up 2 kinds of roles with scope, I'd like `post_submitter` and `commenter`. I guess `poster` and `commenter` could better reinforce the objects structure. If setting up 2 role types is difficult, then I'm up for `submitter` in place of `user`.

zporter replied:
Hmmm if you wanted, an `alias` *could* be setup to have one term alias the `belongs_to` relationship, but since this is a more generic submission model, I would go with a more generic name for the submitter.

zporter replied:
On second thought, `alias` would not work in this instance (sorry, it's been a while since I've Rails'd). The only option to make this work here would be setting 2 `belongs_to` statements. Having 1 sounds better to me.

Other option is to split this model into 2: `Post` and `Comment`. That would allow you to be more specific in roles and functionality.

gingin77 replied:
Hmm... I want to stick with keeping posts and comments in the same model for now and see if I can set up the data table to accommodate nested comments. If that gets too cumbersome, I'll split the model. At the moment, I don't think its super crucial to make a distinction between poster and commenter.

gingin77 replied:
I'm trying to move forward with changing this to `:submitter`, but I have more questions

First - syntax-wise - should the following work?
`belongs_to :submitter, class_name: "User"`

Second, I tried using the syntax above and was getting errors when I tried to rake the seeds file because 'user' wasn't recognized, which I’m guessing is because some of my validations will need to be updated from `user` to `submitter`.

Third, were you thinking that `submitter` would completely replace `user` or be in addition to `user`

zporter replied:
> First - syntax-wise - should the following work?

I think you might need to specify the `foreign_key` as Rails will assume `submitter_id`.  If this is going to be too much of a hassle, then we can pass on doing this. I just thought it'd be nice to have it be a bit more specific as to the role the user was on this table.

> Second...

Might be that, but could also be the foreign key thing that I mentioned above.

> Third, were you thinking that submitter would completely replace user or be in addition to user

The former -- completely replace `user` here with `submitter`.

zporter commented on app/views/submissions/show.html.erb:
```diff
@@ -0,0 +1,38 @@
+    <% end %>
+  <% end %>
+</div>
+<%= form_for(new_comment, url: submission_create_comment_path(post.id), method: :post) do |f| %>
```
I'm tempted to suggest removing the `new_comment` method in favor of inlining with `current_user.submissions.build`.

Also, should just be able to pass `post` into `submission_create_comment_path`.

zporter commented on db/seeds.rb:
```diff
@@ -1,12 +1,29 @@
+  },
+  {
+    user_id: 2,
+    post_id: 1,
```
setting the IDs like this is making a bunch of assumptions about the database -- mainly that it's in a clean state where the sequence to generate the IDs is set to zero. Running the seeds multiple times would yield different results.

To address that:

- Use `exists?` to guard against creating multiples of the same record when re-running the seeds
- Set local variables for the users and posts referenced here to ensure consistency no matter what state the database is in

Does all that make sense?

gingin77 replied:
I've pushed a commit related to this, so I have some kind of solution but wanted to go ahead post this comment since it outlines my process and asks about switching to FactoryBot for seeding the db.

Yes, implementing local variables makes complete sense and I was definitely assuming a blank slate! 

One other thing to consider is that the unique email constraint on the User model blocks the users from being added (based on what I have in my seeds file currently). If the `users` and `submissions` are not added during the same rake, I don't think using local variables would work to pass a `user` to a `submission`. I could switch to using FactoryBot here because I know how to use sequence to change an email address #, which would get around the unique constraint.

Also, I understand the idea of checking whether a record exists but I'm not sure of how to do it. I tried `User.first.exists?` and ended up getting `User.first.nil?` to return an accurate boolean in rails c.

zporter commented on test/factories/submission.rb:
```diff
@@ -0,0 +1,18 @@
+    name "Oliver Sacks"
+    sequence(:email) { |n| "olivers#{n}@email.com" }
+    password "secret"
+  end
```
So this factory file is named `submission`, but it's including a definition for a `user`. I suggest moving this definition to a `test/factories/user.rb` file.

gingin77 replied:
I thought about that before. Would you advise against one main `factories/factories.rb` file?

zporter commented on test/models/submission_test.rb:
```diff
@@ -0,0 +1,50 @@
+    sub_no_user = build(:text_submission, user: nil)
+    assert sub_w_user.valid?
+    assert_not sub_no_user.valid?
+  end
```
when you get to RSpec, check out adding the [`shoulda-matchers` gem](https://github.com/thoughtbot/shoulda-matchers) as well. This provides methods like `it { is_expected.to validate_presence_of(:user_id) }` and many other helpful testing methods.

zporter commented on test/models/submission_test.rb:
```diff
@@ -0,0 +1,50 @@
+  test "submission with text is valid" do
+    sub = build(:text_submission)
+    assert sub.valid?
+  end
```
these 2 tests are just testing that the factory definitions are setup properly. while there's some merit to that, I think I'd rather see specific attributes set in the `build` method and assert that those records are valid.

gingin77 replied:
True.... it made more sense to have these when I first started writing the tests.

gingin77 committed:
Eliminate  method and use Rails  in view instead

gingin77 committed:
Refactor model - include complementary methods to define 2 sub types; rename method  instead of

gingin77 committed:
Small mods in views for submission show and new

gingin77 committed:
Update hash syntax in head tag; update link_to

gingin77 committed:
Refactor error handling by focussing on new submissions form

Errors pertaining to specific fields will display inline based on how  is structured.
If the error message is attached to :base, then only that message will show at the top of the form.
If the error message is associated with a specific attribute, it will display below the field within the form.

Some styling changes that make things good enough for now

gingin77 committed:
Remove jbuilder and coffee-rails gems

zporter commented on app/helpers/application_helper.rb:
```diff
@@ -1,2 +1,9 @@
 module ApplicationHelper
+  def show_errors(object, field_name)
+    if object.errors.any?
+      if !object.errors.messages[field_name].blank?
```
Combine the nested conditional here with an `&&` clause.  Also, slight preference for using `present?` over `!blank?`.

zporter commented on app/views/submissions/new.html.erb:
```diff
@@ -0,0 +1,41 @@
+<div class="show_and_edit">
+  <%= form_for new_post do |post| %>
+    <% if new_post.errors[:base].present? %>
```
the `show_errors` method is already doing this presence check, right?  This seems redundant. Could you move the `div` to the helper method?

gingin77 replied:
"Could you move the div to the helper method?" That's a good idea. The reason I used the `if` statement was because I was getting styled divs in the view when the form was first displayed. If the helper holds the div, then I'd get around that issue.

zporter commented on app/views/submissions/new.html.erb:
```diff
@@ -0,0 +1,41 @@
+      <%= post.label :title %>
+      <%= post.text_field :title %><br/>
+    </div>
+    <% if new_post.errors[:title].present? && !new_post.errors[:base].present? %>
```
why the check for `:base` errors here? Is the idea to not show inline errors if there are base errors?

gingin77 replied:
Exactly!

zporter commented on app/views/submissions/new.html.erb:
```diff
@@ -0,0 +1,41 @@
+      <div class="alert alert-alert inline-alert">
+        <%= show_errors(new_post, :text) %>
+      </div>
+    <% end %>
```
There's a common pattern to rendering these errors. Could the helper do more here to encapsulate that pattern?

gingin77 replied:
Yes, there's definitely a pattern, but I was thinking of some approach from the Rails Views docs might offer a solution. http://guides.rubyonrails.org/action_view_overview.html#partials

Could something like this - 3.2.5 Rendering Collections - work? Or do you have a different approach?

gingin77 committed:
Refactor error message displays - now 2 helpers - one for base; one for inline-field specific messages

gingin77 committed:
Refactor inputs for new submsission form into partials

gingin77 committed:
Update seeds file

Add a conditional to check if Users exist using .nil?
Incorporate local variables to set up belongs_to relationship between submissions and users - in place of usign user_id

gingin77 committed:
Seperate out a factory for users from the submissions factory file

gingin77 committed:
Make 2 tests for validity of submission types more meaningful

gingin77 committed:
Use url_sub predicate method from model in view

gingin77 committed:
Refactor index and show views based on making better use of methods in model

Implement scope to define posts
Refactor related method in controller
Update index and show views to use  predicate method

gingin77 committed:
A single #create method is now responsible for both comment and post submissions

Views and routes were updated to be alligned with this change
Also, some updates to the names of ivars and custom helper_methods
The new_submission method must be memoized for the form errors to correctly display
New validations were written for comments, but I haven't yet gotten these errors to display inline the way I'd like

gingin77 committed:
Minor styling updates related to form layout and error display on forms

gingin77 committed:
A new_comment method that uses an ivar distinct from the submission ivar is needed for a comment submission form to display in the #show template.

If 2 different ivars are not used, then the text content for the existing submission shows up in the text-area of the form for the new submission.
Fix bugs in views where new_post needed to be updated to new_submission or new_comment

gingin77 committed:
Update flash display conditions in layouts file, so that flash can be used to display inline messages on redirects

zporter commented on app/controllers/submissions_controller.rb:
```diff
@@ -2,7 +2,7 @@ class SubmissionsController < ApplicationController
-  helper_method :posts, :post, :new_post, :new_comment
+  helper_method :post, :posts, :new_submission, :new_comment, :parent_id
 
   def new
```
FWIW, you can add `@submission = Submission.new` in this method, replace all instances of `new_submission` with `post` in the views and drop the `new_submission` method in this controller. I'd also recommend aligning the instance variable (`@submission`) with the reader method name (`post`).

gingin77 replied:
Ok, thanks. I'll try that. If I can clarify these things, then I could avoid splitting the model right now.

gingin77 committed:
Update error message for when a blank comment is submitted; update how comment displays in view

zporter commented on app/controllers/submissions_controller.rb:
```diff
@@ -14,26 +14,22 @@ def index
+      else
+        flash[:inline] = @comment.errors.full_messages_for(:text).join(", ")
+        redirect_to submission_path(@comment.post_id)
+      end
```
This conditional can be avoided by adding a `CommentsController` and moving this comment creation logic to the `create` action there.

zporter replied:
I'm wondering if it would be beneficial to also have a `new` view to handle rendering errors...

zporter commented on app/controllers/submissions_controller.rb:
```diff
@@ -47,30 +43,30 @@ def authenticate_to_submit
 
   def new_comment
-    @comment = Submission.new
+    @comment ||= Submission.new
```
I realize I may have said otherwise previously (I honestly can't remember, so sorry if I did), but I wouldn't expect this method to memoize an instance variable. A method prefixed with `new_` should probably return a new instance every time it's invoked.

```ruby
def new_comment
  Submission.new
end
```

Does that make sense?

gingin77 replied:
"A method prefixed with new_ should probably return a new instance every time it's invoked." Yes, that does make sense.... the memoization was first suggested for `new_comment` - https://github.com/gingin77/hackernews_clone/pull/5#pullrequestreview-94148505

zporter replied:
Yea, I'm sure I just commented on the last method I saw with this pattern without thinking too deeply about it. My b.

zporter commented on app/helpers/application_helper.rb:
```diff
@@ -1,13 +1,13 @@
-  def show_errors_on_base(object)
-    if object.errors.any? && object.errors.messages[:base].present?
-      content_tag(:div, object.errors.full_messages_for(:base).join(", "), class: "alert alert-alert")
+  def show_errors_on_base(obj)
```
I'm curious -- why the shortening of the variable name?

gingin77 replied:
I was trying to see if there was logic shared by `show_errors_on_base` and `show_errors_inline` - it was easier to compare when the lines are shorter. Do you think its better to spell out `object`?

zporter replied:
that's cool. I think that code is read more than it is written, so I prefer variable names be verbose enough to reflect what the underlying object is. In cases like this, `object` is fine. I would've also accepted `resource` or `thing`.

zporter commented on app/views/submissions/show.html.erb:
```diff
@@ -31,8 +31,14 @@
 </div>
-<%= form_for(new_comment, url: submission_create_comment_path(post.id), method: :post) do |f| %>
-  <%= f.text_area :text, cols: 60 %>
+<%= form_for(new_comment, url: submissions_path, method: :post) do |comment| %>
```
the reason you see variables like `f` used here is because the underlying variable is an instance of a `FormBuilder`. Using `comment` here might confuse future readers and lead them to thinking it's an instance of a `Comment` class. no action here unless you absolutely want to; I just wanted to throw that out there.

gingin77 replied:
gotcha - that makes sense - I like using `f`

zporter commented on app/views/submissions/show.html.erb:
```diff
@@ -31,8 +31,14 @@
-  <%= f.submit "add comment" %>
+  <%= comment.hidden_field :post_id, :value => parent_id %>
+  <% flash.each do |type, msg| %>
+    <% if type.to_s == "inline" %>
```
hmmm there must be a better way to handle re-rendering someone's comment -- especially if they've typed a really long comment in and some other error happened and they lost their post.

gingin77 replied:
This one was tricky.... It may be another reason why splitting the model could help

gingin77 committed:
Clarify validations again; update test to match custom message

gingin77 committed:
In new comment form builder, use 'f' in place of 'comment'

gingin77 committed:
Rename 'obj' as 'resource'

gingin77 committed:
Fix bugs related the validation method name change; refactor error helpers

gingin77 committed:
Modify seeds file so that Submissions will be created if users are present in AR

gingin77 committed:
Restructure show view using table tags instead of div wrappers
